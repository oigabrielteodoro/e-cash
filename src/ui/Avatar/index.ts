export * from './Avatar'
export * from './NotFoundAvatar'
