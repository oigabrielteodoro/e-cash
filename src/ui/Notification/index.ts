export const NOTIFICATION_EVENT_NAME = 'notification'
export const EXPIRES_NOTIFICATION_IN_SECONDS = 1000 * 3 // 3 seconds

export * from './Notification'
export * from './NotificationContainer'
