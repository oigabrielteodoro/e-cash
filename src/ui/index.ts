export * from './Style'
export * from './Input'
export * from './Button'
export * from './PasswordInput'
export * from './Toast'
export * from './Logo'
export * from './SideBar'
export * from './AppLayout'
export * from './Tooltip'
export * from './Avatar'
export * from './ShimmerEffect'
export * from './Popover'
export * from './Grid'
