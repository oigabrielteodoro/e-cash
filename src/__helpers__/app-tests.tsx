import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
