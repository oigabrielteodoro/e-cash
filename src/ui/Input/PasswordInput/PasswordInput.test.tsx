import React from 'react'
import { useForm } from 'react-hook-form'

import { Input } from 'ui'
import { render, screen, userEvent } from '__helpers__/app-tests'

function MockedComponent() {
  const { register } = useForm()

  return <Input.Password label='Password' {...register('password')} />
}

describe('PasswordInput', () => {
  it('should be able render correctly', async () => {
    render(<MockedComponent />)

    expect(screen.getByLabelText('password')).toBeInTheDocument()
  })

  it('should be able show and hide password', async () => {
    render(<MockedComponent />)

    const input = screen.getByLabelText('password')
    const button = screen.getByRole('button', {
      name: /ControlPasswordType/i,
    })

    expect(input).toBeInTheDocument()
    expect(button).toBeInTheDocument()

    userEvent.type(input, 'John Doe')

    expect(input).toHaveValue('John Doe')

    userEvent.click(button)

    expect(await screen.findByLabelText('password')).toHaveProperty(
      'type',
      'text',
    )

    userEvent.click(
      screen.getByRole('button', {
        name: /ControlPasswordType/i,
      }),
    )

    expect(await screen.findByLabelText('password')).toHaveProperty(
      'type',
      'password',
    )
  })
})
