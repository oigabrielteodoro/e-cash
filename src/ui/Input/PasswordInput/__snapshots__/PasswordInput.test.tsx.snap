// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput should be able match snapshot 1`] = `
.c0 {
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: relative;
}

.c0 label {
  position: absolute;
  font-size: 1.4rem;
  left: 1.2rem;
  top: -0.1rem;
  background: #ffffff;
  padding: 0 0.8rem;
  color: #a1a7ae;
  font-weight: 500;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0.1rem solid #a1a7ae;
  border-radius: 0.5rem;
  margin-top: 0.8rem;
  -webkit-transition: 300ms;
  transition: 300ms;
  width: 100%;
  background: #ffffff;
}

.c1 svg {
  margin-left: auto;
  margin-right: 2rem;
  color: #a1a7ae;
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c1:hover {
  border: 0.1rem solid #3a67e1;
}

.c1:hover svg {
  color: #3a67e1;
}

.c1 input,
.c1 textarea {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  background: transparent;
  padding: 1.8rem 2rem;
  color: #18171d;
  font-size: 1.6rem;
  border-radius: 0.5rem;
}

.c1 input::-webkit-input-placeholder,
.c1 textarea::-webkit-input-placeholder {
  color: #a1a7ae;
}

.c1 input::-moz-placeholder,
.c1 textarea::-moz-placeholder {
  color: #a1a7ae;
}

.c1 input:-ms-input-placeholder,
.c1 textarea:-ms-input-placeholder {
  color: #a1a7ae;
}

.c1 input::placeholder,
.c1 textarea::placeholder {
  color: #a1a7ae;
}

.c1 textarea {
  resize: none;
  padding: 0 2rem;
  margin: 2.4rem 0;
}

.c2 {
  border: 0;
  background: transparent;
}

<div>
  <fieldset
    class="c0"
  >
    <label
      for="password"
    >
      Password
    </label>
    <div
      class="c1"
    >
      <input
        aria-label="password"
        id="password"
        name="password"
        role="textbox"
        type="password"
        value=""
      />
      <button
        aria-label="ControlPasswordType"
        class="c2"
        type="button"
      >
        <svg
          fill="currentColor"
          height="22"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M832 464h-68V240c0-70.7-57.3-128-128-128H388c-70.7 0-128 57.3-128 128v224h-68c-17.7 0-32 14.3-32 32v384c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V496c0-17.7-14.3-32-32-32zM332 240c0-30.9 25.1-56 56-56h248c30.9 0 56 25.1 56 56v224H332V240zm460 600H232V536h560v304zM484 701v53c0 4.4 3.6 8 8 8h40c4.4 0 8-3.6 8-8v-53a48.01 48.01 0 1 0-56 0z"
          />
        </svg>
      </button>
    </div>
  </fieldset>
</div>
`;
